packages/contracts/contracts/AZTEC/BilateralSwap.sol

```diff
@@ -0,0 +1,264 @@
+             * will throw if this is not the case.
+             * @param note the calldata loation of the note
+             **/
+            function validateCommitment(note, k, a) {
+                let gen_order := 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
+                let field_order := 0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47
+                let gammaX := calldataload(add(note, 0x40))
+                let gammaY := calldataload(add(note, 0x60))
+                let sigmaX := calldataload(add(note, 0x80))
+                let sigmaY := calldataload(add(note, 0xa0))
+                if iszero(
+                    and(
+                        and(
+                            and(
+                                eq(mod(a, gen_order), a), // a is modulo generator order?
+                                gt(a, 1)                  // can't be 0 or 1 either!
+                            ),
+                            and(
+                                eq(mod(k, gen_order), k), // k is modulo generator order?
+                                gt(k, 1)                  // and not 0 or 1
+                            )
+                        ),
+                        and(
+                            eq( // y^2 ?= x^3 + 3
+                                addmod(mulmod(
+                                    mulmod(sigmaX, sigmaX, field_order), sigmaX, field_order), 
+                                    3, 
+                                    field_order),
+                                mulmod(sigmaY, sigmaY, field_order)
+                            ),
+                            eq( // y^2 ?= x^3 + 3
+                                addmod(mulmod(
+                                    mulmod(gammaX, gammaX, field_order), 
+                                    gammaX, 
+                                    field_order), 
+                                    3, field_order),
+                                mulmod(gammaY, gammaY, field_order)
+                            )
+                        )
+                    )
+                ) {
+                    mstore(0x00, 400)
+                    revert(0x00, 0x20)
+                }
+            }
+
+            /**
+             * @dev Calculate the keccak256 hash of the commitments for both 
+             * input notes and output notes. This is used both as an input to 
+             * validate the challenge `c` and also to generate pseudorandom relationships
+             * between commitments for different outputNotes, so that we can combine 
+             * them into a single multi-exponentiation for the purposes of validating 
+             * the bilinear pairing relationships.
+             * @param notes calldata location notes
+             * @param n number of notes
+             **/
+            function hashCommitments(notes, n) {
+                for { let i := 0 } lt(i, n) { i := add(i, 0x01) } {
```

Comment:
We know `n=4`, so we can un-roll this loop and use hard-coded offsets to index calldata - saves ~300 gas.

```
calldatacopy(0x300, add(notes, 0x60), 0x80)
calldatacopy(0x380, add(notes, 0x120), 0x80)
calldatacopy(0x400, add(notes, 0x1e0), 0x80)
calldatacopy(0x480, add(notes, 0x2a0), 0x80)
mstore(0x00, keccak256(0x300, 0x200))
```

Reply:
Done